dask/array/core.py

```diff
@@ -1573,9 +1573,479 @@ def __float__(self):
+
+                length -= 1
+
+            len_parsed_indices = len(parsed_indices)
+
+            if ndim and len_parsed_indices > ndim:
+                raise IndexError(
+                    "Invalid indices {} for array with shape {}".format(
+                        parsed_indices, shape
+                    )
+                )
+
+            if len_parsed_indices < ndim:
+                parsed_indices.extend([slice(None)] * (ndim - len_parsed_indices))
+
+            if not ndim and parsed_indices:
+                raise IndexError("Scalar array can only be indexed with () or Ellipsis")
+
+            n_lists = 0
+
+            for i, (index, size) in enumerate(zip(parsed_indices, shape)):
+                is_slice = isinstance(index, slice)
+                if is_slice:
+                    # ------------------------------------------------
+                    # Index is a slice
+                    # ------------------------------------------------
+                    start, stop, step = index.indices(size)
+                    if step < 0 and stop == -1:
+                        stop = None
+
+                    index = slice(start, stop, step)
+
+                elif isinstance(index, (int, np.integer)):
+                    # ------------------------------------------------
+                    # Index is an integer
+                    # ------------------------------------------------
+                    if index < 0:
+                        index += size
+
+                    index = slice(index, index + 1, 1)
+                    is_slice = True
+                else:
+                    n_lists += 1
+                    if n_lists > 1:
+                        raise ValueError(
+                            "Can provide at most one dimension's index as a "
+                            "list of integers or booleans. Got: {}".format(indices)
+                        )
+
+                    convert2positve = True
+                    if getattr(
+                        getattr(index, "dtype", None), "kind", None
+                    ) == "b" or isinstance(index[0], bool):
+                        # --------------------------------------------
+                        # Index is a sequence of booleans, so convert
+                        # the booleans to non-negative integers. We're
+                        # assuming that anything with a dtype
+                        # attribute also has a size attribute.
+                        # --------------------------------------------
+                        if len(index) != size:
+                            raise IndexError(
+                                "Incorrect number ({}) of boolean indices for "
+                                "dimension with size {}: {}".format(
+                                    len(index), size, index
+                                )
+                            )
+                        index = where(index)[0].compute_chunk_sizes()
```

Comment:
This conversion (boolean dtype -> positive indices) seems like potentially not worth the trouble.

`where(cond)` which requires computation at graph construction time (to determine chunk sizes), which is generally best avoided, unlike the three argument form `where(cond, x, y)` can done _lazily_,.

Also, it looks everything in this code path needs to get converted into a `slice()`, which seems very restrictive and thus unlikely to be useful. This is particularly surprising given that dask already supports `x[index] = y` for arbitrary multi-dimensional boolean `index`.

Reply:
Thanks for looking at this, @shoyer - very much appreciated.

Only 1-d integer arrays that _can_ be converted to slices (like `[3, 2, 1]`, or `[1, 3, 5]`) are converted. I have found in the past that this is worth the effort because the numpy \_\_setitem\_\_ is much, much faster with slices, saving more time than we wasted on checking the 1-d array for regularity.

Other 1-d arrays indices (like `[1, 2, 5]`) remain as they are.

I will have a think about how we can process boolean indices without converting them to integers.

Reply:
> I will have a think about how we can process boolean indices without converting them to integers.

... at least, perhaps, until compute time ....

Reply:
@shoyer - apologies, I see now that you had spotted that even _integer_ indices are converted to slices, which is indeed overly restrictive, and caused a bug (https://github.com/dask/dask/pull/7033#issuecomment-757504404). This is bug is now fixed, and I'm thinking about the boolean indices issues.

Reply:
The premature compute from the one-argument where has gone. The move from regular integer sequences to slices has moved to `slicing.setitem` and done chunkwise at compute time (https://github.com/dask/dask/pull/7033/commits/c46ccb2e39bc520f0c8012b823307e823c0631f2)

So, I think there is very little work done now at the time of defining the assignment. The only significant thing that remains is a test for strict monotonicity of integer array indices -  if such an index is a dask array then it is computed at at the time of the assignment definition. It is my hope, however, that we can, in the future, remove the need to this test, and allow other type of integer sequences (liket Array.\_\_getitem\_\_ and numpy do).